perf/reader_test.go

```diff
@@ -390,6 +391,54 @@ func TestPause(t *testing.T) {
 	}
 }
 
+func TestPerfReaderGarbage(t *testing.T) {
+	// This test reproduces a kernel bug that causes garbage reads from the
+	// perf ring as soon as the ring's page boundary has wrapped for the first
+	// time. Sending a 61-byte (unaligned) payload will result in 80 bytes
+	// written to the perf ring:
+	// - perf event header (8 bytes)
+	// - payload (61 + 4 incorrectly added) = 65, aligned to 72
+
+	// The 80 bytes total is already aligned to 64 bits, so won't have any
+	// extra alignment padding when written to the ring. This should cause
+	// the 52nd write to a 4096-byte ring (4096/80 = 51.2) to wrap the page
+	// and return 7 `align(61+4)-4 - 61 = 7` bytes of garbage.
+
+	// This issue might be fixed in later versions of the kernel.
+	sl := 61
+
+	prog, events := mustOutputSamplesProg(t, sl)
+	defer prog.Close()
+	defer events.Close()
+
+	rd, err := NewReader(events, 4096)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer rd.Close()
+
+	for i := 1; i <= 100; i++ {
+		if _, _, err := prog.Test(make([]byte, 14)); err != nil {
+			testutils.SkipIfNotSupported(t, err)
+			t.Fatal(err)
+		}
+
+		record, err := rd.Read()
+		if err != nil {
+			t.Fatal("Can't read samples:", err)
+		}
+
+		// Maintain a slice of zeroes the length of the padding section.
+		nb := make([]byte, len(record.RawSample)-sl)
+
+		// Compare the padding section to the empty slice.
+		if !reflect.DeepEqual(record.RawSample[sl:], nb) {
+			t.Skipf("%d non-zero bytes found in padding section of perf sample %d: %v", len(nb), i, record.RawSample[sl:])
```

Comment:
Hmm, this isn't really a test. What do we gain by including it?

Reply:
I think it's a valuable reproducer, and I hope this behaviour gets fixed in later kernels. As soon as that happens, this test will come up as passed, but on current kernels, it's not an error, hence the skip.

Reply:
I agree it's a good reproducer, but it since it doesn't really have an error case there isn't much point to running it in CI, etc. Even when the bug is fixed the test doesn't increase the quality of the lib / protect us against regressions.

Re maintenance: the reproducer will fail on systems that have a non 4k page size, since in that case the perf ring will be larger than the 4k you ask for. So to really make this work you'd have to try and emulate what the test for lost perf samples does. If you take a look you'll see it's quite involved and I'd rather avoid that.

Reply:
Alright, I've removed it then. It wouldn't protect against regressions, but it would serve as a canary against making assumptions about the buffer length, contents or alignment.

> the reproducer will fail on systems that have a non 4k page size

8k seems to be the lowest common denominator, which would be easy to account for by bumping the cycle count across the board.